File: src/ch04/btree.py
#!/usr/bin/env python3
# btree.py
# A binary tree implementation to refamiliarize with Python
from collections.abc import Callable
from enum import Enum, auto
from typing import Any, Optional, Self, Tuple


class BTreeNode:
    """
    A binary tree node. There's nothing stopping the user from initializing a tree with
    inconsistent types, but don't do it.
    """
    def __init__(self, data: Any=None):
        self.data: Any = data
        self.left_child: Optional[Self] = None
        self.right_child: Optional[Self] = None
    
    def clear(self, which: str="both") -> Tuple[Optional[Self], Optional[Self]]:
        """Removes the specified subtree(s) from this node, returning it as a result"""
        ret = (None, None)
        match which.lower():
            case "left":
                ret = (self.left_child, None)
                self.left_child = None
            case "right":
                ret = (None, self.right_child)
                self.right_child = None
            case "both":
                ret = (self.left_child, self.right_child)
                self.left_child = None
                self.right_child = None
        return ret

    def add(self, node: Self, which: str):
        match which.lower():
            case "left":
                self.left_child = node
            case "right":
                self.right_child = node
            case _:
                raise Exception("Invalid tree child specification")

    def traverse_postorder(self, action: Callable[[Self], None]) -> None:
        class State(Enum):
            LEFT = auto()
            RIGHT = auto()
            ROOT = auto()
        
        state_stack = [State.LEFT]
        node_stack = [self]
        
        while state_stack and node_stack:
            node = node_stack.pop()
            state = state_stack.pop()
            
            match state:
                case State.LEFT:
                    node_stack.append(node)
                    state_stack.append(State.RIGHT)

                    if node.left_child:
                        node_stack.append(node.left_child)
                        state_stack.append(State.LEFT)

                case State.RIGHT:
                    node_stack.append(node)
                    state_stack.append(State.ROOT)

                    if node.right_child:
                        node_stack.append(node.right_child)
                        state_stack.append(State.LEFT)

                case State.ROOT:
                    action(node)

    def __str__(self):
            return str(self.data)

    def init_BTree(vals: list[int] | list[str]) -> Self:
        assert(vals)
        vals.reverse()
        print(vals)
        root_node = BTreeNode(vals.pop())
        
        nodes = [root_node]

        def fill(node: BTreeNode):
            # Fill via level order
            left = vals.pop()
            node.left_child = BTreeNode(left)
            nodes.insert(0, node.left_child)

            right = vals.pop()
            node.right_child = BTreeNode(right)
            nodes.insert(0, node.right_child)
        
        while(vals and nodes):
            curr_node = nodes.pop()
            try:
                fill(curr_node)
            except Exception:
                continue

        return root_node
            



if __name__ == "__main__":

    # Test the function
    def printNode(node: BTreeNode) -> None:
        try:
            print(f"{node.data}", end=" ")
        except Exception:
            print("Could not print the node's data.")


    btree = BTreeNode.init_BTree([x for x in range(5)])
    btree.traverse_postorder(printNode)
    print("")

File: src/ch04/btree.cpp
// btree.cpp
// See implementation result at: https://godbolt.org/z/fP5zcKbPs

#include <cassert>
#include <functional>
#include <iostream>
#include <memory>
#include <stdexcept>
#include <string>
#include <vector>

/// Think: Ownership. There is only one way to access a binary tree node: through its parent. 
/// The parent-child relationship implies ownership (parent owns children).
template<typename T>
class BTreeNode
{
public:
    enum class Direction {left, right};
    
    // CONSTRUCTORS
    explicit BTreeNode(T data)
        : d_data(data) {}   

    BTreeNode(T data, std::shared_ptr<BTreeNode> subtree, Direction dir)
        : d_data(data)
    {
        switch(dir)
        {
            case Direction::lefT:
                d_left_child = subtree;
                break;
            case Direction::right:
                d_right_child = subtree;
                break;
            default:
                throw std::runtime_error("Impossible variant\n");
        }
    }

    BTreeNode(T data, std::shared_ptr<BTreeNode> left_tree, std::shared_ptr<BTreeNode> right_tree)
        : d_data(data), d_left_child(left_tree), d_right_child(right_tree)
    {}

    // ACCESSORS
    auto data() const -> T const& { return d_data; }

    // MANIPULATORS

    /// Traverse the tree, performing the input (non-modifying) operation 
    static void traverse_postorder(
        std::shared_ptr<BTreeNode> tree,
        std::function< void( std::shared_ptr<BTreeNode<T>> )> action
    );

private:
    T d_data;
    std::optional< std::shared_ptr< BTreeNode>> d_left_child {};
    std::optional< std::shared_ptr< BTreeNode>> d_right_child {};

    /// Track which part of the node is being visited.
    enum class TraversalState { root, left, right };

};

int main() 
{
    auto b6 = std::make_shared<BTreeNode<int>>(6);
    auto b5 = std::make_shared<BTreeNode<int>>(5);
    auto b4 = std::make_shared<BTreeNode<int>>(4);
    auto b3 = std::make_shared<BTreeNode<int>>(3);

    auto b2 = std::make_shared<BTreeNode<int>>(2, b5, b6);
    auto b1 = std::make_shared<BTreeNode<int>>(1, b3, b4);

    auto b0 = std::make_shared<BTreeNode<int>>(0, b1, b2);

    std::cout << "Post-order Traversal: ";
    BTreeNode<int>::traverse_postorder(
        b0, 
        [](std::shared_ptr<BTreeNode<int>> node) { 
            std::cout << node->data() << " "; 
            return;
        }
    );
}


//------------------------------------------------------------------------------
template<typename T>
// 
void BTreeNode<T>::traverse_postorder(
    std::shared_ptr<BTreeNode> tree,
    std::function< void( std::shared_ptr<BTreeNode<T>> )> action
)
{
    using State = BTreeNode<T>::TraversalState;
    // Init Tracker variables
    std::vector<State> states { State::left };
    std::vector< std::shared_ptr< BTreeNode<T>>> nodes { tree };

    while (!nodes.empty() && !states.empty())
    {
        auto node = nodes.back();
        // nodes.pop_back();
        auto state = states.back();
        states.pop_back();

        switch(state)
        {
            case State::left:
                states.push_back(State::right);
                if (node->d_left_child)
                {
                    nodes.push_back(node->d_left_child.value());
                    states.push_back(State::left);
                }
                break;
            case State::right:
                states.push_back(State::root);
                if (node->d_right_child)
                {
                    nodes.push_back(node->d_right_child.value());
                    states.push_back(State::left);
                }
                break;
            case State::root:
                action(node);
                // remove the node
                nodes.pop_back();
                break;
            default:
                throw std::runtime_error("Unknown traversal state");
        }
    }
}
